lib/rack/builder.rb

```diff
@@ -34,7 +34,7 @@ class Builder
     def self.parse_file(config, opts = Server::Options.new)
       options = {}
       if config =~ /\.ru$/
-        cfgfile = ::File.read(config)
+        cfgfile = ::File.read(config, mode: 'r:bom|utf-8')
```

Comment:
Does this introduce a unstated requirement that `config.ru` is UTF-8?

Reply:
Seems that way, doesn’t it? I’ll concoct a utf-16 config.ru or perhaps another encoding and see what happens. 

Thanks for giving this PR a look!

Reply:
This change does not introduce a new requirement. Instead it makes a previous requirement explicit.

When not told the encoding of the file, `::File.read` assumes the content is UTF-8 and returns a string encoded as UTF-8 even when file is encoded as UTF-16. This fails because when the string is `eval`ed, the first null byte is interpreted as EOF, which leads to a `SyntaxError`.

I'm not sure it's possible to infer the encoding of the `config.ru` file by looking at it, in general. I believe it is preferable to raise a bit earlier with a more reasonable error, as my proposed change would.

tested with Ruby 2.5.1

Reply:
Sorry to disappoint, but `::File.read` doesn't assume UTF-8 unless your `LC_CTYPE` environment variable or `Encoding.default_external` say so. So I don't think that this is backwards compatible.

Reply:
Thanks for the clarification on that point, @manveru.

Unfortunately, changing my `Encoding.default_external` to match the content of my `config.ru` only leads to different problems.

First, the simple call to `File.read` fails with "ArgumentError: ASCII incompatible encoding needs binmode".

Second, `File.read` with `mode: 'rb'`, doesn't assume the content matches my default external encoding and provides `ASCII-8BIT` instead.

Finally, if I explicitly declare the content type for `File.read` with `mode: 'rb:utf-16le'`, it fails to successfully concatenate with the literal strings in the next method.
`Encoding::CompatibilityError: incompatible character encodings: UTF-8 and UTF-16LE`

Additional pointers on these issues are welcome.

Reply:
I don't think we should support non-ASCII compatible encodings for source code. There is no way to set `LC_*` to UTF-16.
Also try to run the following file:

    # Encoding: UTF-16LE
    p 'hi'.encoding

And see that Ruby fails with `UTF-16LE is not ASCII compatible (ArgumentError)`.

Reply:
Thanks again, @manveru. I think I've got it now. 

I was able to create a CP1252 `config.ru` with some accented characters and it worked as expected. Assuming `config.ru` is UTF-8 would be a breaking change.

----

Regarding the issue this PR is meant to solve, how do you feel about a solution like:

- if the string encoding is UTF-8 after reading the file,
- strip the BOM with `String` methods or a regexp

?